src/zarr/core/metadata/dtype.py

```diff
@@ -0,0 +1,576 @@
+    name: ClassVar[str]
+    dtype_cls: ClassVar[type[TDType]]  # this class will create a numpy dtype
+    kind: ClassVar[DataTypeFlavor]
+    default_value: TScalar
+
+    def __init_subclass__(cls) -> None:
+        # Subclasses will bind the first generic type parameter to an attribute of the class
+        # TODO: wrap this in some *very informative* error handling
+        generic_args = get_args(get_original_bases(cls)[0])
+        cls.dtype_cls = generic_args[0]
+        return super().__init_subclass__()
+
+    def to_dict(self) -> dict[str, JSON]:
+        return {"name": self.name}
+
+    def cast_value(self: Self, value: object, *, endianness: Endianness | None = None) -> TScalar:
+        return cast(np.generic, self.to_dtype(endianness=endianness).type(value))
+
+    @classmethod
+    @abstractmethod
+    def from_dtype(cls: type[Self], dtype: TDType) -> Self:
+        raise NotImplementedError
+
+    def to_dtype(self: Self, *, endianness: Endianness | None = None) -> TDType:
+        endian_str = endianness_to_numpy_str(endianness)
+        return self.dtype_cls().newbyteorder(endian_str)
+
+    @abstractmethod
+    def to_json_value(self, data: np.generic, *, zarr_format: ZarrFormat) -> JSON:
+        """
+        Convert a single value to JSON-serializable format. Depends on the zarr format.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def from_json_value(
+        self: Self, data: JSON, *, zarr_format: ZarrFormat, endianness: Endianness | None = None
+    ) -> TScalar:
+        """
+        Read a JSON-serializable value as a numpy scalar
+        """
+        raise NotImplementedError
+
+
+@dataclass(frozen=True, kw_only=True)
+class Bool(DTypeWrapper[np.dtypes.BoolDType, np.bool_]):
+    name = "bool"
+    kind = "boolean"
+    default_value = np.False_
+
+    @classmethod
+    def from_dtype(cls, dtype: np.dtypes.BoolDType) -> Self:
+        return cls()
+
+    def to_json_value(self, data: np.generic, zarr_format: ZarrFormat) -> bool:
+        return bool(data)
+
+    def from_json_value(
+        self, data: JSON, *, zarr_format: ZarrFormat, endianness: Endianness | None = None
+    ) -> np.bool_:
+        if check_json_bool(data):
+            return self.to_dtype(endianness=endianness).type(data)
+        raise TypeError(f"Invalid type: {data}. Expected a boolean.")
+
+
+class IntWrapperBase(DTypeWrapper[TDType, TScalar]):
+    kind = "numeric"
+
+    @classmethod
+    def from_dtype(cls, dtype: TDType) -> Self:
+        return cls()
+
+    def to_json_value(self, data: np.generic, zarr_format: ZarrFormat) -> int:
+        return int(data)
+
+    def from_json_value(
+        self, data: JSON, *, zarr_format: ZarrFormat, endianness: Endianness | None = None
+    ) -> TScalar:
+        if check_json_int(data):
+            return self.to_dtype(endianness=endianness).type(data)
+        raise TypeError(f"Invalid type: {data}. Expected an integer.")
```

Comment:
I use inheritance for dtypes like the integers, which really only differ in their concrete dtype + scalar types.